### PR TITLE
chore(api): mypy enabled on drivers package

### PR DESCRIPTION
**Pull request by amitlissack**
# Overview

Removed mypy exceptions for init, algorithms, and test-algorithms.

See #8511

# Changelog

Fixed (?) typing issues.

# Review requests

# Risk assessment

None

### CODE REVIEW

File: api/src/opentrons/drivers/serial_communication.py [hunk 1]
```diff
@@ -67,7 +75,7 @@ def _write_to_device_and_return(cmd, ack, device_connection, tag=None):
 
     encoded_write = cmd.encode()
     encoded_ack = ack.encode()
-    log.debug(f"{tag}: Write -> {encoded_write}")
+    log.debug(f"{tag}: Write -> {encoded_write!r}")
```
Comment (Laura-Danielle) [inline api/src/opentrons/drivers/serial_communication.py:78]: What's the point of the `!r` syntax here?

**Review by Laura-Danielle — APPROVED** [review]
One question, and it also seems like you might have some files from [your other PR](https://github.com/Opentrons/opentrons/pull/8514) in here for some reason?

Otherwise looks great.

File: api/src/opentrons/drivers/serial_communication.py (hunk 1 above)
Comment (amitlissack) [reply api/src/opentrons/drivers/serial_communication.py:78]: For `encoded_write` being a `bytes` object.

File: api/src/opentrons/drivers/asyncio/communication/async_serial.py [hunk 2]
```diff
@@ -33,16 +33,29 @@ async def create(
...
-                baudrate=baud_rate,
-                timeout=timeout,
-                write_timeout=write_timeout,
-            ),
+            executor,
+            _create_serial,
+            port,
+            baud_rate,
+            timeout,
+            write_timeout,
```
Comment (mcous) [inline api/src/opentrons/drivers/asyncio/communication/async_serial.py]: Could we avoid ditching the use of keyword args here?

```python
        serial = await loop.run_in_executor(
            executor=executor,
            func=functools.partial(
                _create_serial,
                url=port,
                baudrate=baud_rate,
                rto=timeout,
                wto=write_timeout,
            )
        )
```

File: api/src/opentrons/drivers/asyncio/communication/async_serial.py [hunk 3]
```diff
@@ -85,7 +98,7 @@ async def read_until(self, match: bytes, timeout: Optional[float] = None) -> byt
         """
         with self._timeout_override("timeout", timeout):
             return await self._loop.run_in_executor(
-                executor=self._executor, func=lambda: self._serial.read_until(match)
+                self._executor, self._serial.read_until, match
```
Comment (mcous) [inline api/src/opentrons/drivers/asyncio/communication/async_serial.py]: Same comment. Would prefer functools.partial and keyword args

File: api/src/opentrons/drivers/asyncio/communication/async_serial.py [hunk 4]
```diff
@@ -147,10 +160,12 @@ async def is_open(self) -> bool:
...
         Returns: boolean
         """
-        return self._serial.is_open
+        return self._serial.is_open is True
 
     @contextlib.contextmanager
-    def _timeout_override(self, timeout_property: str, timeout: Optional[float]):
+    def _timeout_override(
+        self, timeout_property: str, timeout: Optional[float]
+    ) -> Generator[None, None, None]:
```
Comment (mcous) [inline api/src/opentrons/drivers/asyncio/communication/async_serial.py:158]: Any thoughts on typing this as `Iterator[None]` vs `Generator`? Just curious, not asking for a change. [Mypy docs](https://mypy.readthedocs.io/en/stable/kinds_of_types.html#generators) say:

> A basic generator that only yields values can be annotated as having a return type of either `Iterator[YieldType]` or `Iterable[YieldType]`. For example:
>
> ```python
> def squares(n: int) -> Iterator[int]:
>     for i in range(n):
>         yield i * i
> ```

File: api/src/opentrons/drivers/rpi_drivers/gpio.py [hunk 5]
```diff
@@ -47,16 +47,16 @@ def board_rev(self) -> BoardRevision:
...
 
-    def _request_line(self, pin: GPIOPin, request_type) -> gpiod.Line:
+    def _request_line(self, pin: GPIOPin, request_type: int) -> gpiod.Line:
         name = pin.name
         offset = pin.by_board_rev(self.board_rev)
 
         line = self.chip.get_line(offset)
 
-        def _retry_request_line(retries: int = 0):
+        def _retry_request_line(retries: int = 0):  # type: ignore
```
Comment (mcous) [inline api/src/opentrons/drivers/rpi_drivers/gpio.py]: Can we make sure all added `type: ignore` comments are scoped to the specific violation?

File: api/src/opentrons/drivers/rpi_drivers/gpio.py [hunk 6]
```diff
@@ -160,39 +160,39 @@ def set_button_light(
...
         if on:
             self.set_high(gpio_group.audio_enable)
         else:
             self.set_low(gpio_group.audio_enable)
 
-    def _read(self, input_pin: GPIOPin):
+    def _read(self, input_pin: GPIOPin) -> int:
         try:
-            return self.lines[input_pin.name].get_value()
+            return self.lines[input_pin.name].get_value()  # type: ignore
```
Comment (mcous) [inline api/src/opentrons/drivers/rpi_drivers/gpio.py]: Unscoped `type: ignore`

File: api/src/opentrons/drivers/rpi_drivers/gpio_simulator.py [hunk 7]
```diff
@@ -27,48 +27,50 @@ def board_rev(self) -> BoardRevision:
...
 
     def _initialize(self) -> Dict[str, int]:
-        lines = {}
-        for pin in gpio_group.pins:
-            lines[pin.name] = pin.by_board_rev(self.board_rev)
+        name_pin_num = (
+            (pin.name, pin.by_board_rev(self.board_rev)) for pin in gpio_group.pins
+        )
+        # Dict comprehension eliminating None values.
+        lines = {k: v for (k, v) in name_pin_num if v is not None}
```
Comment (mcous) [inline api/src/opentrons/drivers/rpi_drivers/gpio_simulator.py]: Can we change this back to a for-in loop? Personally I find previous version more readable

```suggestion
        lines = {}
        for pin in gpio_group.pins:
            line = pin.by_board_rev(self.board_rev)
            if line:
                lines[pin.name]
```

File: api/src/opentrons/drivers/rpi_drivers/types.py [hunk 8]
```diff
@@ -165,18 +174,20 @@ class GPIOGroup:
     def __init__(self, pins: List[GPIOPin]):
         self.pins = pins
 
-    def __getattr__(self, item):
-        return next(filter(lambda x: x.name is item, self.pins), None)
+    def __getattr__(self, item: str) -> GPIOPin:
+        res = next(filter(lambda x: x.name is item, self.pins), None)
+        assert res, f"Failed to find GPIOPin named: {item}"
```
Comment (mcous) [inline api/src/opentrons/drivers/rpi_drivers/types.py:179]: That `None` from `next` seems pretty intentional, is this `assert` safe to add?

File: api/src/opentrons/drivers/serial_communication.py [hunk 9]
```diff
@@ -1,10 +1,13 @@
-from typing import List
+from typing import List, Optional, Generator
 
 import serial  # type: ignore
+from serial import Serial
 from serial.tools import list_ports  # type: ignore
 import contextlib
 import logging
 
+from serial.tools.list_ports_common import ListPortInfo  # type: ignore
```
Comment (mcous) [inline api/src/opentrons/drivers/serial_communication.py]: Unscoped `type: ignore`

File: api/src/opentrons/drivers/serial_communication.py [hunk 10]
```diff
@@ -25,15 +28,18 @@ def get_ports_by_name(device_name: str) -> List[serial.Serial]:
...
         if d.vid == vid:
             return d[0]
+    return None
 
 
 @contextlib.contextmanager
-def serial_with_temp_timeout(serial_connection, timeout):
+def serial_with_temp_timeout(
+    serial_connection: Serial, timeout: float
+) -> Generator[Serial, None, None]:
```
Comment (mcous) [inline api/src/opentrons/drivers/serial_communication.py]: Same `Generator` vs `Iterator` question

File: api/src/opentrons/drivers/smoothie_drivers/driver_3_0.py [hunk 11]
```diff
@@ -530,7 +530,9 @@ def steps_per_mm(self) -> Dict[str, float]:
         return self._steps_per_mm
 
     @contextlib.asynccontextmanager
-    async def restore_speed(self, value: Union[float, str]):
+    async def restore_speed(
+        self, value: Union[float, str]
+    ) -> AsyncGenerator[None, None]:
```
Comment (mcous) [inline api/src/opentrons/drivers/smoothie_drivers/driver_3_0.py]: AsyncGenerator vs AsyncIterator

**Review by mcous — COMMENTED** [review]
Thanks for taking this on! Some comments and questions

File: api/src/opentrons/drivers/rpi_drivers/types.py (hunk 8 above)
Comment (amitlissack) [reply api/src/opentrons/drivers/rpi_drivers/types.py:179]: The usage of this method is that it returns `GPIOPin`. Not `Optional`. At least this assert will make the error easier to track.

File: api/src/opentrons/drivers/rpi_drivers/gpio_simulator.py (hunk 7 above)
Comment (amitlissack) [reply api/src/opentrons/drivers/rpi_drivers/gpio_simulator.py]: Ok

File: api/src/opentrons/drivers/asyncio/communication/async_serial.py (hunk 2 above)
Comment (amitlissack) [reply api/src/opentrons/drivers/asyncio/communication/async_serial.py]: good idea

File: api/src/opentrons/drivers/asyncio/communication/async_serial.py (hunk 4 above)
Comment (amitlissack) [reply api/src/opentrons/drivers/asyncio/communication/async_serial.py:158]: Much nicer

File: api/src/opentrons/drivers/asyncio/communication/async_serial.py (hunk 3 above)
Comment (amitlissack) [reply api/src/opentrons/drivers/asyncio/communication/async_serial.py]: Much nicer!

**Review by mcous — APPROVED** [review]
🦖

**Pull request merged**